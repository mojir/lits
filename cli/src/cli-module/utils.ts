export type Module = Record<string, unknown>
export type GetFsModule = (modulePath: string[]) => Module
