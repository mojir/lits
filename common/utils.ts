import type { UnknownRecord } from '../src/interface'
import { isRegularExpression } from '../src/typeGuards/lits'
import { isLitsFunction } from '../src/typeGuards/litsFunction'

export function stringifyValue(value: unknown, html: boolean): string {
  const gt = html ? '&gt;' : '>'
  const lt = html ? '&lt;' : '<'
  if (isLitsFunction(value)) {
    if (value.functionType === 'Builtin')
      return `${lt}builtin function ${value.n}${gt}`
    else
      return `${lt}function ${(value as unknown as UnknownRecord).n ?? '\u03BB'}${gt}`
  }
  if (value === null)
    return 'null'

  if (typeof value === 'object' && value instanceof Error)
    return value.toString()

  if (typeof value === 'object' && value instanceof RegExp)
    return `${value}`

  if (value === Number.POSITIVE_INFINITY)
    return `${Number.POSITIVE_INFINITY}`

  if (value === Number.NEGATIVE_INFINITY)
    return `${Number.NEGATIVE_INFINITY}`

  if (typeof value === 'number' && Number.isNaN(value))
    return 'NaN'

  if (isRegularExpression(value))
    return `/${value.s}/${value.f}`

  if (typeof value === 'string')
    return `"${value}"`

  return JSON.stringify(value, null, 2)
}

export function findAllOccurrences(input: string, pattern: RegExp): Set<string> {
  const matches = [...input.matchAll(pattern)]
  return new Set(matches.map(match => match[0]))
}
