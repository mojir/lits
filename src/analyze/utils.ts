import type { UnresolvedSymbol, UnresolvedSymbols } from '.'

export function joinAnalyzeResults(...results: UnresolvedSymbols[]): UnresolvedSymbols {
  const result = new Set<UnresolvedSymbol>()
  for (const symbols of results)
    symbols.forEach(symbol => result.add(symbol))

  return result
}

export function addAnalyzeResults(target: UnresolvedSymbols, source: UnresolvedSymbols): void {
  source.forEach(symbol => target.add(symbol))
}

export function combinate<T>(arrays: T[][]): T[][] {
  return arrays.reduce((acc: T[][], curr) => {
    return acc.flatMap(a => curr.map(c => [...a, c]))
  }, [[]])
}
