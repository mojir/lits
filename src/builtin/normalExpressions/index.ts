import type { BuiltinNormalExpression, BuiltinNormalExpressions } from '../interface'
import type { Any } from '../../interface'
import type { FunctionReference } from '../../../reference'
import type { NormalExpressionName } from '../../../reference/api'
import { bitwiseNormalExpression } from './categories/bitwise'
import { collectionNormalExpression } from './categories/collection'
import { arrayNormalExpression } from './categories/array'
import { sequenceNormalExpression } from './categories/sequence'
import { mathNormalExpression } from './categories/math'
import { miscNormalExpression } from './categories/misc'
import { assertNormalExpression } from './categories/assert'
import { objectNormalExpression } from './categories/object'
import { predicatesNormalExpression } from './categories/predicates'
import { regexpNormalExpression } from './categories/regexp'
import { stringNormalExpression } from './categories/string'
import { functionalNormalExpression } from './categories/functional'
import { gridNormalExpression } from './categories/namespaces/grid'
import { vectorNormalExpression } from './categories/namespaces/vector'
import { linearAlgebraNormalExpression } from './categories/namespaces/linearAlgebra'
import { matrixNormalExpression } from './categories/namespaces/matrix'
import { combinatoricalNormalExpression } from './categories/namespaces/numberTheory'
import { randomNormalExpression } from './categories/namespaces/random'
import { getMetaNormalExpression } from './categories/meta'

const normalExpressionReference: Record<string, FunctionReference> = {}

export function setNormalExpressionReference(reference: Record<NormalExpressionName, FunctionReference>) {
  Object.assign(normalExpressionReference, reference)
}

const expressions: BuiltinNormalExpressions = {
  ...bitwiseNormalExpression,
  ...collectionNormalExpression,
  ...arrayNormalExpression,
  ...sequenceNormalExpression,
  ...mathNormalExpression,
  ...getMetaNormalExpression(normalExpressionReference),
  ...miscNormalExpression,
  ...assertNormalExpression,
  ...objectNormalExpression,
  ...predicatesNormalExpression,
  ...regexpNormalExpression,
  ...stringNormalExpression,
  ...functionalNormalExpression,
  ...vectorNormalExpression,
  ...linearAlgebraNormalExpression,
  ...gridNormalExpression,
  ...matrixNormalExpression,
  ...combinatoricalNormalExpression,
  ...randomNormalExpression,
}

Object.entries(expressions).forEach(([name, expression]) => {
  expression.name = name
})

const aliases: BuiltinNormalExpressions = {}

Object.values(expressions).forEach((normalExpression) => {
  normalExpression.aliases?.forEach((alias) => {
    aliases[alias] = normalExpression
  })
})

export const normalExpressions: BuiltinNormalExpressions = {
  ...expressions,
  ...aliases,
}

export const normalExpressionTypes: Record<string, number> = {}
export const allNormalExpressions: BuiltinNormalExpression<Any>[] = []

Object.entries(normalExpressions).forEach(([key, value], index) => {
  normalExpressionTypes[key] = index
  allNormalExpressions.push(value)
})
