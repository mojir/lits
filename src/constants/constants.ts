export const NodeTypes = {
  Number: 1,
  String: 2,
  NormalExpression: 3,
  SpecialExpression: 4,
  UserDefinedSymbol: 5,
  NormalBuiltinSymbol: 6,
  SpecialBuiltinSymbol: 7,
  ReservedSymbol: 8,
  Binding: 9,
  Spread: 10,
} as const

const NodeTypesSet = new Set(Object.values(NodeTypes))

export type NodeType = typeof NodeTypes[keyof typeof NodeTypes]

export function getNodeTypeName(type: NodeType): keyof typeof NodeTypes {
  return Object.keys(NodeTypes).find(key => NodeTypes[key as keyof typeof NodeTypes] === type) as keyof typeof NodeTypes
}

// TODO, is this needed?
export function isNodeType(type: unknown): type is NodeType {
  return typeof type === 'number' && NodeTypesSet.has(type as NodeType)
}

const functionTypes = [
  'UserDefined',
  'Partial',
  'Comp',
  'Constantly',
  'Juxt',
  'Complement',
  'EveryPred',
  'SomePred',
  'Fnull',
  'Builtin',
  'NativeJsFunction',
] as const

const functionTypeSet = new Set(functionTypes)

export type FunctionType = typeof functionTypes[number]

export function isFunctionType(type: unknown): type is FunctionType {
  return typeof type === 'string' && functionTypeSet.has(type as FunctionType)
}
