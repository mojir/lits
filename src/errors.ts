import { getCodeMarker } from '../src/utils/debug/getCodeMarker'
import type { Arr } from './interface'
import type { SourceCodeInfo } from './tokenizer/token'

function getLitsErrorMessage(message: string, sourceCodeInfo?: SourceCodeInfo) {
  const filePathLine = sourceCodeInfo?.filePath ? `\n${sourceCodeInfo.filePath}` : ''
  const codeLine = sourceCodeInfo?.code ? `\n${sourceCodeInfo.code}` : ''
  const codeMarker = sourceCodeInfo && codeLine ? `\n${getCodeMarker(sourceCodeInfo)}` : ''
  return `${message}${filePathLine}${codeLine}${codeMarker}`
}

export class RecurSignal extends Error {
  public params: Arr
  constructor(params: Arr) {
    super(`recur, params: ${params}`)
    Object.setPrototypeOf(this, RecurSignal.prototype)
    this.name = 'RecurSignal'
    this.params = params
  }
}

export class LitsError extends Error {
  public readonly sourceCodeInfo?: SourceCodeInfo
  public readonly shortMessage: string
  constructor(message: string | Error, sourceCodeInfo: SourceCodeInfo | undefined) {
    if (message instanceof Error)
      message = `${message.name}${message.message}`

    super(getLitsErrorMessage(message, sourceCodeInfo))
    this.shortMessage = message
    this.sourceCodeInfo = sourceCodeInfo
    Object.setPrototypeOf(this, LitsError.prototype)
    this.name = 'LitsError'
  }

  public getCodeMarker(): string | undefined {
    return this.sourceCodeInfo && getCodeMarker(this.sourceCodeInfo)
  }
}

export class NotAFunctionError extends LitsError {
  constructor(fn: unknown, sourceCodeInfo?: SourceCodeInfo) {
    const message = `Expected function, got ${fn}.`
    super(message, sourceCodeInfo)
    Object.setPrototypeOf(this, NotAFunctionError.prototype)
    this.name = 'NotAFunctionError'
  }
}

export class UserDefinedError extends LitsError {
  public userMessage: string
  constructor(message: string, sourceCodeInfo?: SourceCodeInfo) {
    super(message, sourceCodeInfo)
    this.userMessage = message
    Object.setPrototypeOf(this, UserDefinedError.prototype)
    this.name = 'UserDefinedError'
  }
}

export class AssertionError extends LitsError {
  constructor(message: string | Error, sourceCodeInfo?: SourceCodeInfo) {
    super(message, sourceCodeInfo)
    Object.setPrototypeOf(this, AssertionError.prototype)
    this.name = 'AssertionError'
  }
}

export class UndefinedSymbolError extends LitsError {
  public symbol: string
  constructor(symbolName: string, sourceCodeInfo?: SourceCodeInfo) {
    const message = `Undefined symbol '${symbolName}'.`
    super(message, sourceCodeInfo)
    this.symbol = symbolName
    Object.setPrototypeOf(this, UndefinedSymbolError.prototype)
    this.name = 'UndefinedSymbolError'
  }
}

export function isLitsError(error: unknown): error is LitsError {
  return error instanceof LitsError
}
