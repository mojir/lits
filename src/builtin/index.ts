import type { Builtin } from './interface'
import { allNormalExpressions, normalExpressions } from './normalExpressions'
import { andSpecialExpression } from './specialExpressions/and'
import { condSpecialExpression } from './specialExpressions/cond'
import { switchSpecialExpression } from './specialExpressions/switch'
import { definedSpecialExpression } from './specialExpressions/defined'
import { defSpecialExpression } from './specialExpressions/def'
import { doSpecialExpression } from './specialExpressions/block'
import { lambdaSpecialExpression } from './specialExpressions/functions'
import { ifSpecialExpression } from './specialExpressions/if'
import { unlessSpecialExpression } from './specialExpressions/unless'
import { letSpecialExpression } from './specialExpressions/let'
import { loopSpecialExpression } from './specialExpressions/loop'
import { doseqSpecialExpression, forSpecialExpression } from './specialExpressions/loops'
import { orSpecialExpression } from './specialExpressions/or'
import { qqSpecialExpression } from './specialExpressions/qq'
import { recurSpecialExpression } from './specialExpressions/recur'
import { throwSpecialExpression } from './specialExpressions/throw'
import { trySpecialExpression } from './specialExpressions/try'
import { arraySpecialExpression } from './specialExpressions/array'
import { objectSpecialExpression } from './specialExpressions/object'
import { specialExpressionTypes } from './specialExpressionTypes'

export const specialExpressions = [
  qqSpecialExpression,
  andSpecialExpression,
  orSpecialExpression,
  arraySpecialExpression,
  condSpecialExpression,
  defSpecialExpression,
  definedSpecialExpression,
  // defnSpecialExpression,
  doSpecialExpression,
  doseqSpecialExpression,
  lambdaSpecialExpression,
  forSpecialExpression,
  // functionSpecialExpression,
  ifSpecialExpression,
  letSpecialExpression,
  loopSpecialExpression,
  objectSpecialExpression,
  recurSpecialExpression,
  switchSpecialExpression,
  throwSpecialExpression,
  trySpecialExpression,
  unlessSpecialExpression,
] as const

export type SpecialExpressions = typeof specialExpressions
export type SpecialExpression = SpecialExpressions[number]
export type SpecialExpressionName = keyof typeof specialExpressionTypes
export type CommonSpecialExpressionType = [
  | typeof specialExpressionTypes['??']
  | typeof specialExpressionTypes['&&']
  | typeof specialExpressionTypes['cond']
  | typeof specialExpressionTypes['switch']
  | typeof specialExpressionTypes['defined?']
  | typeof specialExpressionTypes['block']
  | typeof specialExpressionTypes['if']
  | typeof specialExpressionTypes['unless']
  | typeof specialExpressionTypes['||']
  | typeof specialExpressionTypes['throw']
  | typeof specialExpressionTypes['array']
  | typeof specialExpressionTypes['object'],
]

export type SpecialExpressionType = typeof specialExpressionTypes[SpecialExpressionName]

export const builtin: Builtin = {
  normalExpressions,
  specialExpressions,
  allNormalExpressions,
}

export const normalExpressionKeys = Object.keys(normalExpressions)
export const specialExpressionKeys = Object.keys(specialExpressionTypes)
export const specialExpressionNameSet = new Set(specialExpressionKeys)

// TODO, remove
// console.log('builtin', [...specialExpressionKeys, ...normalExpressionKeys].length)
