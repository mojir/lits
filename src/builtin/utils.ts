import { LitsError } from '../errors'
import type { ContextStack } from '../evaluator/ContextStack'
import type { AstNode, BindingNode } from '../parser/types'
import type { SourceCodeInfo } from '../tokenizer/token'
import { isReservedSymbol } from '../tokenizer/reservedNames'
import type { Builtin } from './interface'
import type { SpecialExpressionName } from '.'

export type Arity = number | { min: number }

export interface FunctionOverload {
  as: FunctionArguments
  a: Arity
  b: AstNode[]
}

export interface FunctionArguments {
  m: string[]
  r?: string
  b: BindingNode[]
}

export function assertNameNotDefined<T>(
  name: T,
  contextStack: ContextStack,
  builtin: Builtin,
  sourceCodeInfo?: SourceCodeInfo,
): asserts name is T {
  if (typeof name !== 'string')
    return

  if (builtin.specialExpressions[name as SpecialExpressionName])
    throw new LitsError(`Cannot define variable ${name}, it's a special expression.`, sourceCodeInfo)

  if (builtin.normalExpressions[name])
    throw new LitsError(`Cannot define variable ${name}, it's a builtin function.`, sourceCodeInfo)

  if (isReservedSymbol(name))
    throw new LitsError(`Cannot define variable ${name}, it's a reserved name.`, sourceCodeInfo)

  if (contextStack.globalContext[name])
    throw new LitsError(`Name already defined "${name}".`, sourceCodeInfo)
}
