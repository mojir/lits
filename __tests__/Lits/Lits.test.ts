import { beforeEach, describe, expect, it } from 'vitest'
import { UndefinedSymbolError } from '../../src/errors'
import { Cache } from '../../src/Lits/Cache'
import type { LazyValue } from '../../src/Lits/Lits'
import { Lits } from '../../src/Lits/Lits'
import { assertLitsFunction } from '../../src/typeGuards/litsFunction'
import { FUNCTION_SYMBOL } from '../../src/utils/symbols'
import type { Ast, NormalExpressionNodeWithName, UserDefinedFunction } from '../../src/parser/types'
import { NodeTypes } from '../../src/constants/constants'
import { normalExpressionTypes } from '../../src/builtin/normalExpressions'

describe('all tests', () => {
  describe('tEST', () => {
    let lits: Lits
    beforeEach(() => {
      lits = new Lits({ debug: true, astCacheSize: 0 })
    })
    it('without params', () => {
      const fn = lits.run('-> $1 + $2')
      assertLitsFunction(fn)
      expect(lits.apply(fn, [2, 3])).toBe(5)
    })
    it('with empty params', () => {
      const fn = lits.run('-> $1 + $2')
      assertLitsFunction(fn)
      expect(lits.apply(fn, [2, 3], {})).toBe(5)
    })

    it('with params', () => {
      const fn = lits.run('-> $1 + $2 + x')
      assertLitsFunction(fn)
      expect(lits.apply(fn, [2, 3], { contexts: [{ x: { value: 1 } }] })).toBe(6)
    })
  })

  describe('lazy host values as function', () => {
    it('that it works', () => {
      const lits = new Lits()
      const lazyHostValues: Record<string, LazyValue> = {
        x: {
          read: () => 42,
        },
        foo: {
          read: () => ({
            [FUNCTION_SYMBOL]: true,
            name: undefined,
            functionType: 'UserDefined',
            evaluatedfunction: [[], [[NodeTypes.Number, 42]], {}],
          } satisfies UserDefinedFunction),
        },
      }

      expect(lits.run('x', { lazyValues: lazyHostValues })).toBe(42)
      expect(lits.run('foo()', { lazyValues: lazyHostValues })).toBe(42)
      expect(lits.run('z', { lazyValues: { z: { read: () => 12 } } })).toBe(12)
    })
  })

  describe('runtime info', () => {
    it('getRuntimeInfo().', () => {
      const lits = new Lits()
      expect(lits.getRuntimeInfo()).toMatchSnapshot()
    })
    it('getRuntimeInfo() with ast cache > 0', () => {
      const lits = new Lits({ astCacheSize: 10 })
      expect(lits.getRuntimeInfo()).toMatchSnapshot()
    })
    it('getRuntimeInfo() with ast cache = 0', () => {
      const lits = new Lits({ astCacheSize: 0 })
      expect(lits.getRuntimeInfo()).toMatchSnapshot()
    })
  })

  describe('context', () => {
    let lits: Lits
    beforeEach(() => {
      lits = new Lits({ debug: true })
    })
    it('a function.', () => {
      lits = new Lits({ astCacheSize: 10 })
      const contexts = [lits.context('export function tripple(x) x * 3 end;')]
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
    })
    it('a function with ast.', () => {
      lits = new Lits({ astCacheSize: 10 })
      const parseResult = lits.parse(lits.tokenize('export function tripple(x) x * 3 end;'))
      const contexts = [lits.context(parseResult)]
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
    })

    it('a function - no cache', () => {
      lits = new Lits({ debug: true })
      const contexts = [lits.context('export function tripple(x) x * 3 end;', {})]
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
      expect(lits.run('tripple(10)', { contexts })).toBe(30)
    })

    it('a function - initial cache', () => {
      const initialCache: Record<string, Ast> = {
        '2 ^ 4': {
          hasDebugData: false,
          body: [
            [NodeTypes.NormalExpression, [[NodeTypes.NormalBuiltinSymbol, normalExpressionTypes['^'] as number], [[NodeTypes.Number, 2], [NodeTypes.Number, 2]]]] satisfies NormalExpressionNodeWithName,
          ],
        },
      }
      lits = new Lits({ astCacheSize: 10, initialCache })
      expect(lits.run('2 ^ 2')).toBe(4)
      expect(lits.run('2 ^ 4')).toBe(4)
    })

    it('a variable.', () => {
      const contexts = [lits.context('export let magicNumber := 42;')]
      expect(lits.run('magicNumber', { contexts })).toBe(42)
    })

    it('a variable - again.', () => {
      const contexts = [
        lits.context(`
    export function zip?(input) boolean(match(input, #"^\\d{5}$")) end;
    export let NAME_LENGTH := 100;
    `),
      ]
      expect(lits.run('NAME_LENGTH', { contexts })).toBe(100)
    })

    it('a function with a built in normal expression name', () => {
      expect(() => lits.context('function inc(x) x + 1 end')).toThrow()
    })

    it('a function with a built in special expression name', () => {
      expect(() => lits.context('function and(x) x + 1 end')).toThrow()
    })

    it('a variable twice', () => {
      const contexts = [lits.context('export let magicNumber := 42; export function getMagic() 42 end;')]
      lits.context('export let magicNumber := 42; export function getMagic() 42 end;', { contexts })
    })

    it('more than one', () => {
      const contexts = [lits.context('export function tripple(x) x * 3 end;'), lits.context('export let magicNumber := 42;')]
      expect(lits.run('tripple(magicNumber)', { contexts })).toBe(126)
    })
  })

  function ast(n: number): Ast {
    return {
      hasDebugData: false,
      body: [[NodeTypes.Number, n]],
    }
  }

  describe('cache', () => {
    it('cannot set same key twice', () => {
      const cache = new Cache(10)
      cache.set('a', ast(1))
      expect(() => cache.set('a', ast(2))).toThrow()
    })

    it('getContent', () => {
      const cache = new Cache(10)
      cache.set('a', ast(1))
      cache.set('b', ast(2))
      expect(cache.getContent()).toEqual({
        a: ast(1),
        b: ast(2),
      })
    })
    it('getContent (null)', () => {
      const cache = new Cache(null)
      cache.set('a', ast(1))
      cache.set('b', ast(2))
      expect(cache.getContent()).toEqual({
        a: ast(1),
        b: ast(2),
      })
    })

    it('max cache size must be at least 1', () => {
      expect(() => new Cache(-1)).toThrow()
      expect(() => new Cache(0)).toThrow()
      expect(() => new Cache(0.1)).not.toThrow()
      expect(() => new Cache(1)).not.toThrow()
    })

    it('add an entry.', () => {
      const cache = new Cache(10)
      expect(cache.size).toBe(0)
      cache.set('a', ast(1))
      expect(cache.size).toBe(1)
      expect(cache.get('a')).toEqual(ast(1))
      expect(cache.has('a')).toBe(true)
    })

    it('clear cache.', () => {
      const cache = new Cache(10)
      cache.set('a', ast(1))
      cache.set('b', ast(2))
      cache.set('c', ast(3))
      expect(cache.size).toBe(3)
      cache.clear()
      expect(cache.size).toBe(0)
    })

    it('add an entry - cacheSize = 1', () => {
      const cache = new Cache(1)
      expect(cache.size).toBe(0)
      cache.set('a', ast(1))
      expect(cache.size).toBe(1)
      expect(cache.get('a')).toEqual(ast(1))
    })
    it('maxSize.', () => {
      const cache = new Cache(1)
      cache.set('a', ast(1))
      expect(cache.get('a')).toEqual(ast(1))
      cache.set('b', ast(2))
      expect(cache.size).toBe(1)
      expect(cache.get('a')).toBeUndefined()
      expect(cache.has('a')).toBe(false)
      expect(cache.get('b')).toEqual(ast(2))
      expect(cache.has('b')).toBe(true)
    })
  })

  describe('regressions', () => {
    let lits: Lits
    beforeEach(() => {
      lits = new Lits({ debug: true })
    })
    it.skip('sourceCodeInfo', () => {
      try {
        lits.run('let n := 3 write!(n)') // Missing semi
      }
      catch (error) {
      // eslint-disable-next-line ts/no-unsafe-member-access
        expect((error as any).sourceCodeInfo.position.line).toBe(3)
        // eslint-disable-next-line ts/no-unsafe-member-access
        expect((error as any).sourceCodeInfo.position.column).toBe(7)
      }
    })
    it('name not recognized', () => {
      expect(() => lits.run('asd()')).toThrowError(UndefinedSymbolError)
      expect(() => lits.run('asd')).toThrowError(UndefinedSymbolError)
    })

    it('unexpected argument', () => {
      try {
        lits.run('1 + + 2')
      }
      catch (error) {
      // eslint-disable-next-line ts/no-unsafe-assignment
        const anyError = error as any
        // eslint-disable-next-line ts/no-unsafe-member-access
        expect(anyError.sourceCodeInfo.position.line).toBe(1)
        // eslint-disable-next-line ts/no-unsafe-member-access
        expect(anyError.sourceCodeInfo.position.column).toBe(7)
      }
    })

    it('shoud handle double quoted in strings', () => {
    // You need to escape double quote with a backslash
      expect(lits.run('"\\""')).toBe('"')
      // You need to escape backslash with a backslash if it is at the end of the string
      expect(lits.run('"\\\\"')).toBe('\\')
      // You need to escape backslash with a backslash if it is followed by a double quote
      expect(lits.run('"\\"\\\\\\""')).toBe('"\\"')
      // Backslash before normal character is returning the character itself
      expect(lits.run('"\\abc"')).toBe('abc')
    })
  })
})
