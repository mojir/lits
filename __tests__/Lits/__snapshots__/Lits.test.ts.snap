// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`all tests > runtime info > getRuntimeInfo() with ast cache = 0 1`] = `
{
  "astCache": null,
  "astCacheSize": 0,
  "debug": false,
}
`;

exports[`all tests > runtime info > getRuntimeInfo() with ast cache > 0 1`] = `
{
  "astCache": Cache {
    "_size": 0,
    "cache": {},
    "firstEntry": undefined,
    "lastEntry": undefined,
    "maxSize": 10,
  },
  "astCacheSize": 10,
  "debug": false,
}
`;

exports[`all tests > runtime info > getRuntimeInfo(). 1`] = `
{
  "astCache": null,
  "astCacheSize": null,
  "debug": false,
}
`;

exports[`runtime info > getRuntimeInfo() with ast cache = 0 1`] = `
{
  "astCache": null,
  "astCacheSize": 0,
  "debug": false,
}
`;

exports[`runtime info > getRuntimeInfo() with ast cache > 0 1`] = `
{
  "astCache": Cache {
    "_size": 0,
    "cache": {},
    "firstEntry": undefined,
    "lastEntry": undefined,
    "maxSize": 10,
  },
  "astCacheSize": 10,
  "debug": false,
}
`;

exports[`runtime info > getRuntimeInfo(). 1`] = `
{
  "astCache": null,
  "astCacheSize": null,
  "debug": false,
}
`;
